File: javascript/leetcode/129. Sum Root to Leaf Numbers.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function (root) {
  const add = (root, sum) => {
    if (!root) return 0;

    let total = sum * 10 + root.val;

    if (!root.right && !root.left) return total;

    return add(root.left, total) + add(root.right, total);
  };

  return add(root, 0);
};

File: c++/leetcode/129. Sum Root to Leaf Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  int add(TreeNode *, int);
  int sumNumbers(TreeNode *root)
  {
    return add(root, 0);
  }
};

int Solution::add(TreeNode *root, int sum)
{
  if (!root)
    return 0;

  int total = (sum * 10) + root->val;

  if (!root->left && !root->right)
    return total;

  return add(root->left, total) + add(root->right, total);
}